File: Suffix Structures.cpp
#include <bits/stdc++.h>
using namespace std;

int main() {
    string s, t;
    cin >> s >> t;

    vector<int> cnt_s(26, 0), cnt_t(26, 0);
    for (int i = 0; i < s.size(); i++) {
        cnt_s[s[i] - 'a']++;
    }

    for (int i = 0; i < t.size(); i++) {
        cnt_t[t[i] - 'a']++;
    }

    for (int i = 0; i < 26; i++) {
        if (cnt_s[i] < cnt_t[i]) {
            cout << "need tree";
            return 0;
        } 
    }

    bool automaton = true;
    int j = 0;
    for (int i = 0; i < s.size(); i++) {
        if (s[i] == t[j]) {
            j++;
        }
    }

    if (j < t.size()) {
        automaton = false;
    }

    if (s.size() == t.size()) {
        cout << "array";
    }
    else if (automaton) {
        cout << "automaton";
    }
    else {
        cout << "both";
    }

    return 0;
}

File: algorithms/python/intermediate/dynamic.array_and_string/Suffix Structures.py
def cnt_char(s):
    char = [0] * 26
    for c in s:
        char[ord(c) - ord('a')] += 1
    return char


def check_type(s, t):
    char_s = cnt_char(s)
    char_t = cnt_char(t)

    for i in range(26):
        if char_t[i] > char_s[i]:
            return "need tree"

    if len(t) == len(s):
        return "array"  # only need shuffle position from s

    j = 0
    for i in range(len(s)):
        if s[i] == t[j]:
            j += 1
            if j == len(t):
                return "automaton"  # only need delete any position to change from s to t
    return "both"


s = input()
t = input()
print(check_type(s, t))